### PR TITLE
Fix xcvrd to support 400G ZR optic

### PR DESCRIPTION
**Pull request by abohanyang**
Fix Xcvrd code to support 400G ZR optic

#### Description
While bringing the 400G ZR module, I found an issue where some 400G ZR optics may take more than 60 secs to finish up the Datapath initialization. This is a problem because the xcvrd uses a hardcoded 60 secs Datapath initialization timeout for all CMIS modules and apparently, this 60 secs timeout is not sufficient for the 400G ZR module.
The CMIS spec has Datapath init/de-init timeout values defined in Byte 144. We can use these values from the EEPROM instead of presenting any hardcoded values.

#### Motivation and Context
This is needed to bring up 400G ZR module.

#### How Has This Been Tested?
These changes are tested against arista_7800r3a_36d2_lc and arista_7280cr3_32p4 platforms.

### CODE REVIEW

**Comment by abohanyang** [comment]
@arlakshm

**Comment by prgeor** [comment]
@abohanyang can you please check if on_port_update_event() is triggered even after this PR [https://github.com/sonic-net/sonic-platform-daemons/pull/285](https://github.com/sonic-net/sonic-platform-daemons/pull/285). If so please share the syslog file.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1640]: where is 'stopping_even' used in this function? can you point me to the caller of this function where its passing 'stopping_even' argument?

**Review by prgeor — CHANGES_REQUESTED** [review]

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py:1640]: Hi Prince - thank you for taking a look. On line 1756, 
"port_change_event_handler = functools.partial(self.on_port_config_change, stopping_event)"

adds an argument, stopping_event, to this handler function. when port_change_event_handler gets called in read_port_config_change (xcvrd_utilities/port_mapping.py), it actually calls self.on_port_config_change(stopping_event,port_change_event).

Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py]: I guess I know why you are confused about my change. I just notice that I did not patch the code correctly. You can check my latest diff.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [reply sonic-xcvrd/xcvrd/xcvrd.py:926]: Following PRs were raised:
https://github.com/sonic-net/sonic-platform-daemons/issues/290
https://github.com/sonic-net/sonic-platform-daemons/issues/294

Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py:926]: To read CMIS dp state duration, I create another PR in sonic-platform-common (https://github.com/sonic-net/sonic-platform-common/pull/312).

Comment (shyam77git) [reply sonic-xcvrd/xcvrd/xcvrd.py:926]: @abohanyang  - How is #312 (above) different from #290 ?

Comment (shyam77git) [reply sonic-xcvrd/xcvrd/xcvrd.py:926]: Hi @abohanyang 
Is #312 sufficient to take care of what's mentioned in #290 or anything more required?
Note that #294 to take care for xcvrd side changes.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [inline sonic-xcvrd/xcvrd/xcvrd.py:1542]: @abohanyang : get_cmis_expired() - is this a temporary function defined to validate this fix - #293?

practically,  get_datapath_init_duration() and  get_datapath_init_duration() and get_datapath_deinit_duration() to be invoked instead. 
Would you be taking care of this as part of this PR? or shall we take care of this in xcvrd.py via #294 ?

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [inline sonic-xcvrd/xcvrd/xcvrd.py:1963]: @abohanyang 
Per the description, "on_port_config_change takes three arguments, not two. This issue is found when the port breakout mode changes"
Have a Q: 'stopping event' is passed as an argument but not used/consumed?
How is this related to port breakout changes? what's missing/expected?
the platform/board you are finding & validating 'port breakout change' is PHY-less (i.e. does not have Ext PHY (gearbox))?

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py]: No. @312 contains the implementation of API, get_datapath_init_duration. I just commit more changes in this PR.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py]: Yes. I just commit more changes in this PR to call get_datapath_init_duration and get the expiration time.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py:1959]: Yes, this stopping_event is passed and is not used at all. I actually don't know why we have this stopping_event. I guess this stopping_event will be used to handle some error cases.

My understanding is that this on_port_config_change function is triggered during the port breakout change since the old port is removed (PORT_REMOVE) from the port_mapping and the new port is added (PORT_ADD) to the port_mapping. 

My platform is a PHY-less platform. DNX NIF is connected to the OSFP/QSFP-DD front panel directly.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [reply sonic-xcvrd/xcvrd/xcvrd.py]: > No. 312 contains the implementation of API, get_datapath_init_duration. I just commit more changes in this PR.

Thanks for the clarification. 
This PR now is going to take care of git issue #293 - I'm reviewing this further

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [reply sonic-xcvrd/xcvrd/xcvrd.py]: > Yes. I just commit more changes in this PR to call get_datapath_init_duration and get the expiration time.

OK.
@abohanyang : can you please attach UT cases, logs for this case (400G ZR optics FSM working via xcrd.py as orchestrator with these timeouts all the way to interface in 'operational up' state)

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [inline sonic-xcvrd/xcvrd/xcvrd.py:1480]: this block is to handle CMIS DP_DEINIT state.
Instead of get_cmis_dp_init_duration(api) there should be function get_cmis_dp_deinit_duration(api) defined and invoked here?

DP_PATH_DEINIT_DURATION is defined at the bit range 4-8 vs DP_PATH_INIT_DURATION is defined at bit range 0-4 per CMIS spec (cmis.py), so values could be different.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (shyam77git) [reply sonic-xcvrd/xcvrd/xcvrd.py:1959]: > Yes, this stopping_event is passed and is not used at all. I actually don't know why we have this stopping_event. I guess this stopping_event will be used to handle some error cases.
> 

If unclear, can we avoid adding this change to this PR?

> My understanding is that this on_port_config_change function is triggered during the port breakout change since the old port is removed (PORT_REMOVE) from the port_mapping and the new port is added (PORT_ADD) to the port_mapping.
> 

Would you be validating port breakout as part of this PR? If not, then IMO, this should be handled in separate PR

> My platform is a PHY-less platform. DNX NIF is connected to the OSFP/QSFP-DD front panel directly.
Thanks for clarifying.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1516]: TODO not required now

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1537]: TODO not required now

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1518]: We actually dont need this because at this stage DpInit is not triggered, it just we selected the application so we should check the DpInitPending bit at Page 11h Byte 235

**Review by prgeor — APPROVED** [review]

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py:1959]: Okay. I can create a separate PR to fix this.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py:1518]: I can add code to check DpInitPending additionally but I am not sure if we can simply remove this line. If the module never sets the pending bits somehow, then it will get stuck in the CMIS_STATE_AP_CONF state.

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py]: [log.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9709709/log.txt)

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1096]: can you specify return value duration msec/sec/min?

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1226]: I think we should check if DpInitPending is set for ALL host lanes and return TRUE. Return False if any of the lanes has DpInitPending=0


image

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1545]: if any of the lanes has DPInitPending=0 i.e if this fn call return False then we should retry re-init

Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py]: I have updated the code based on our discussion.

**Comment by abohanyang** [comment]
Hi @shyam77git - The code is pending review. I am waiting for @prgeor  comments/approval on my latest commit.

**Comment by prgeor** [comment]
> Fix Xcvrd code to support 400G DR and 400G ZR optics
> 
> #### Description
> 1. Increase the initialization timeout value for the 400GBASE-ZR module as 400GBASE-ZR may take more than 60 sec to come up. Based on the spec,  ZR may take up to 200 sec to become ready.
> 2. on_port_update_event is triggered when 400G optics enter the ready state. To avoid the infinite initialization loop, in on_port_update_event, call force_cmis_reinit only when port configuration changes.
> 3. on_port_config_change takes three arguments, not two. This issue is found when the port breakout mode changes.
> 
> #### Motivation and Context
> * Item 1 and Item 2  are needed to bring up 400G ZR/DR optics.
> * Item 3 fixes a potential issue during dynamic breakout mode change.
> 
> #### How Has This Been Tested?
> * These changes are tested against arista_7800r3a_36d2_lc and arista_7280cr3_32p4 platforms.

@abohanyang can you check the description. I think you are addressing only #1

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1493]: can you log the deinit duration?

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd.py:1556]: can you please log the dp_init pending duration?

File: sonic-xcvrd/xcvrd/xcvrd.py
Comment (abohanyang) [reply sonic-xcvrd/xcvrd/xcvrd.py]: Sure. I just update the code and the summary of this PR.

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@yxieca can you help cherry-pick to 202205?

**Pull request merged**
